### ISSUE TITLE
Unloading a plugin after an exception & unrolling UI stack can crash the game

### ISSUE THREAD

**Issue by XertroV** [openplanet-nl/openplanet-issues#532]
I'm not exactly sure what causes this yet (probably something to do with resources not being cleaned up / undone properly), but I have run into it a few times with E++ somewhat recently.

If there's an exception after `UI::Begin` but before `UI::End`, E++ will crash ofc and so I need to reload. For everything I've tried so far, I will always get a CTD when unloading E++. (No difference is made by any of: in/out of editor, unloading but not reloading, stopping script engine. I haven't tried toggling it off first then reloading, but I doubt it'd help)

This only seems to matter if there's an unrolling dangling UI stack msg in the logs. Having an exception in Render isn't enough.

Not sure how to replicate it yet (besides deliberately doing it via E++)

**Comment by codecat** [comment]
I've found a potentially related crash which I've fixed now. In the next build, let me know if this is resolved.

**Comment by XertroV** [comment]
Will do, thanks!

**Comment by codecat** [comment]
Have you tried with 1.26.30?

**Comment by codecat** [comment]
Assuming fixed.

**Comment by XertroV** [comment]
It's not fixed. (using 1.26.33)

**Comment by codecat** [comment]
In that case, I will need some reproduce steps.

**Comment by XertroV** [comment]
Yeah okay.

Might be related to dlls not unloading then attempting to be loaded?

**Comment by XertroV** [comment]
Hmm... Or patches or hooks or any of a variety of things.

**Comment by XertroV** [comment]
The crash occurs from `EXCEPTION_BREAKPOINT` (80000003) inside a call to `ucrtbase.free`.

![image](https://github.com/user-attachments/assets/94a7ef99-b2f8-4f21-96b6-3fbbc563035c)

To help locate this, here's bytecode corresponding to the stack trace. I realize this might not be that useful, but on the off chance it is:

#### func1: calls ucrtbase.free

```
Openplanet.dll.text+147E30 - 40 53                 - push rbx
Openplanet.dll.text+147E32 - 48 83 EC 20           - sub rsp,20 { 32 }
Openplanet.dll.text+147E36 - 48 8D 05 83857700     - lea rax,[Openplanet.dll.rdata+2D3C0] { (7FFA02C98E30) }
Openplanet.dll.text+147E3D - 48 8B D9              - mov rbx,rcx
Openplanet.dll.text+147E40 - 48 89 01              - mov [rcx],rax
Openplanet.dll.text+147E43 - F6 C2 01              - test dl,01 { 1 }
Openplanet.dll.text+147E46 - 74 0A                 - je Openplanet.dll.text+147E52
Openplanet.dll.text+147E48 - BA 08000000           - mov edx,00000008 { 8 }
Openplanet.dll.text+147E4D - E8 167A7200           - call Openplanet.dll.text+86F868 { calls ucrtbase.free }
Openplanet.dll.text+147E52 - 48 8B C3              - mov rax,rbx
Openplanet.dll.text+147E55 - 48 83 C4 20           - add rsp,20 { 32 }
Openplanet.dll.text+147E59 - 5B                    - pop rbx
```

#### func2: calls func1

```
Openplanet.dll.text+1810AB - 49 8D 40 FF           - lea rax,[r8-01]
Openplanet.dll.text+1810AF - 48 3B D0              - cmp rdx,rax
Openplanet.dll.text+1810B2 - 73 17                 - jae Openplanet.dll.text+1810CB
Openplanet.dll.text+1810B4 - 48 8B 01              - mov rax,[rcx]
Openplanet.dll.text+1810B7 - 4C 2B C2              - sub r8,rdx
Openplanet.dll.text+1810BA - 49 C1 E0 03           - shl r8,03 { 3 }
Openplanet.dll.text+1810BE - 48 8D 0C D0           - lea rcx,[rax+rdx*8]
Openplanet.dll.text+1810C2 - 48 8D 51 08           - lea rdx,[rcx+08]
Openplanet.dll.text+1810C6 - E8 23F76E00           - call Openplanet.dll.text+8707EE { ->->VCRUNTIME140.memmove }
Openplanet.dll.text+1810CB - 48 FF 4F 08           - dec qword ptr [rdi+08]
Openplanet.dll.text+1810CF - 48 85 DB              - test rbx,rbx
Openplanet.dll.text+1810D2 - 74 0D                 - je Openplanet.dll.text+1810E1
Openplanet.dll.text+1810D4 - 48 8B 03              - mov rax,[rbx]
Openplanet.dll.text+1810D7 - BA 01000000           - mov edx,00000001 { 1 }
Openplanet.dll.text+1810DC - 48 8B CB              - mov rcx,rbx
Openplanet.dll.text+1810DF - FF 10                 - call qword ptr [rax] { calls func1 }
Openplanet.dll.text+1810E1 - 48 8B 5C 24 40        - mov rbx,[rsp+40]
Openplanet.dll.text+1810E6 - B0 01                 - mov al,01 { 1 }
Openplanet.dll.text+1810E8 - 48 83 C4 20           - add rsp,20 { 32 }
Openplanet.dll.text+1810EC - 5F                    - pop rdi
Openplanet.dll.text+1810ED - C3                    - ret 
```

#### func3

```
Openplanet.dll.text+85D603 - 4D 89 7D 00           - mov [r13+00],r15
Openplanet.dll.text+85D607 - 49 83 C5 08           - add r13,08 { 8 }
Openplanet.dll.text+85D60B - 48 83 C6 08           - add rsi,08 { 8 }
Openplanet.dll.text+85D60F - 41 83 EC 08           - sub r12d,08 { 8 }
Openplanet.dll.text+85D613 - 75 EB                 - jne Openplanet.dll.text+85D600
Openplanet.dll.text+85D615 - 49 83 FB 00           - cmp r11,00 { 0 }
Openplanet.dll.text+85D619 - 74 17                 - je Openplanet.dll.text+85D632
Openplanet.dll.text+85D61B - 66 41 0F12 03         - movlpd xmm0,[r11]
Openplanet.dll.text+85D620 - 66 41 0F12 4B 08      - movlpd xmm1,[r11+08]
Openplanet.dll.text+85D626 - 66 41 0F12 53 10      - movlpd xmm2,[r11+10]
Openplanet.dll.text+85D62C - 66 41 0F12 5B 18      - movlpd xmm3,[r11+18]
Openplanet.dll.text+85D632 - 41 FF D6              - call r14 { calls func2 }
Openplanet.dll.text+85D635 - 48 8B E5              - mov rsp,rbp
Openplanet.dll.text+85D638 - 48 83 C4 50           - add rsp,50 { 80 }
Openplanet.dll.text+85D63C - 5B                    - pop rbx
Openplanet.dll.text+85D63D - 41 5F                 - pop r15
Openplanet.dll.text+85D63F - 41 5E                 - pop r14
Openplanet.dll.text+85D641 - 41 5D                 - pop r13
Openplanet.dll.text+85D643 - 41 5C                 - pop r12
Openplanet.dll.text+85D645 - 5F                    - pop rdi
Openplanet.dll.text+85D646 - 41 5B                 - pop r11
Openplanet.dll.text+85D648 - 5E                    - pop rsi
Openplanet.dll.text+85D649 - 5D                    - pop rbp
Openplanet.dll.text+85D64A - C3                    - ret 
```

**Comment by codecat** [comment]
Thanks, I'll take a look.

**Comment by codecat** [comment]
I don't have enough information to be sure exactly what is causing this, but it is related to a plugin resource; either custom allocated memory using `Dev::Allocate` or a hook using `Dev::Hook`. Are you doing either of those?

**Comment by XertroV** [comment]
Yeah it's E++ so I'm using both. I can try avoiding calls to `Dev::Allocate` though and see if that fixes it.

**Comment by codecat** [comment]
I don't know which one it is, but it's one of those. This is a crash on plugin unload when it's trying to clear the memory for the resource.

**Comment by XertroV** [comment]
Would it be possible to call the OnDestroyed() method of the plugin when the UI stack gets unrolled? That would probably fix it for me.

**Comment by codecat** [comment]
Huh? Plugins don't get destroyed when they're marked as unstable, so that wouldn't make sense.

**Comment by XertroV** [comment]
Well they (plugins) don't run after being marked unstable, and OnDestroyed is the cleanup function and it isn't called if the plugin is marked as unstable.
So unstable plugins aren't cleaned up.

Is there a way to mark a plugin as unstable besides the UI stuff? If not, 'unstable' plugins could still have their other (non-render) callbacks called.

**Comment by codecat** [comment]
I need to think about this a bit. `OnDestroyed` is currently blocked by the "regular" set of checks, which also includes whether the plugin is enabled or not. As such, `OnDestroyed` will be skipped on unload if a plugin is disabled before unload. I've made a new issue here: #685
